src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,133 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        if (index < size) {
+            T[] tempElementData = (T[]) new Object[elementData.length];
+            System.arraycopy(elementData, 0, tempElementData, 0, index + 1);
+            tempElementData[index] = value;
+            System.arraycopy(elementData, index, tempElementData, index + 1, size - index);
+            elementData = tempElementData;
+            size++;
+            return;
+        }
+        if (index == size) {
+            elementData[size] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add list");
+        }
+        while (elementData.length < size + list.size()) {
+            elementData = grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isInvalid(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " is invalid");
+        }
```

Comment:
Let's extract all common logic to a dedicated method (including throwing an exception). You only use isInvalid in that specific scenario